### ISSUE TITLE
clarify when componentDidUpdate fires

### ISSUE THREAD

**Issue by mbrowne** [reactjs/reactjs.org#1135]
Re-posting this issue from facebook/react:
https://github.com/facebook/react/issues/2796

> https://reactjs.org/docs/react-component.html#componentdidupdate is unclear on when it is called with regards to the render function.

> The docs state that it's "Invoked immediately after updating occurs" but it is unclear what is meant with updating in this case: it can refer to either the component as a whole updating (meaning post-render), or it can refer to setState() having been processed (which might mean pre-render).

> It would be good to explicitly state when this function fires in its documentation (even if somewhere else on the site this is described in text or graphic form already).

There was a PR for this that was merged at one point: https://github.com/facebook/react/pull/2801

But the current website does not include this change. I believe the language from the above PR is still correct and would be helpful to people since it's more specific than what's there currently.

**Comment by bmueller-sykes** [comment]
I have a similar question: does `componentDidUpdate()` fire after *every* render, or only renders that result in an actual change to the DOM?  e.g. if the virtual DOM is the same as the actual DOM, then componentDidUpdate() still fire?

**Comment by mbrowne** [comment]
The suggested change to the docs was:
> Invoked immediately after the component's updates are flushed to the DOM. This method is not called for the initial render.

This seems to not be quite correct however. For example, the DOM for this component should never change (at least in theory):

```
class Test extends React.Component {
    componentDidUpdate(prevProps) {
        console.log('componentUpdated', prevProps, this.props)
    }

    render() {
        return <div>hi</div>
    }
}
```

...but `componentDidUpdate()` still fires if there are prop or state changes. e.g. this prop change (showing only partial code):

```
        setTimeout(() => {
            this.setState({ foo: 2 })
        })
...
<Test foo={this.state.foo} />
```

I am not an expert on the `componentDidUpdate`; all I know is that it would be helpful for the docs to be more explicit about when it fires. Hopefully someone with more knowledge -- or someone who takes the time to research it -- can open a PR to address this. (But FYI, there is a long backlog of un-reviewed PRs in this repo [including one of my own, unrelated to this issue], so unfortunately the docs probably won't be updated anytime soon even if someone does create a PR.)

**Comment by mbrowne** [comment]
I edited my above comment to show a full code example.

**Comment by bmueller-sykes** [comment]
Right, now it makes more sense. And yeah, I think it is the case that `componentDidUpdate()` fires after *every* `render()`, even though it's not super clear in the docs.
